From 6838a259344bbc3ceca0d1309641ec9f3d59ba51 Mon Sep 17 00:00:00 2001
From: P Praneesh <quic_ppranees@quicinc.com>
Date: Thu, 1 Sep 2022 17:49:58 +0530
Subject: [PATCH] ath12k: allocate few more MSIs for monitor ring

Currently monitor ring mask is shared among dp rx and dp tx interrupts.
Sharing of monitor interrupts with Tx/Rx handler causes performance
overhead. Hence dedicate 4 new MSI to process the rx monitor ring data.

Signed-off-by: P Praneesh <quic_ppranees@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/hw.c  | 11 +++++++++--
 drivers/net/wireless/ath/ath12k/hw.h  |  2 +-
 drivers/net/wireless/ath/ath12k/pci.c | 12 +++++++++---
 3 files changed, 19 insertions(+), 6 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/hw.c
+++ b/drivers/net/wireless/ath/ath12k/hw.c
@@ -308,12 +308,14 @@ const struct ath12k_hw_ring_mask ath12k_
 		ATH12K_TX_RING_MASK_1,
 		ATH12K_TX_RING_MASK_2,
 		ATH12K_TX_RING_MASK_3,
+		0, 0, 0, 0, 0, 0, 0, 0,
 	},
 	.rx_mon_dest = {
-		0, 0, 0,
+		0, 0, 0, 0, 0, 0, 0, 0,
 		ATH12K_RX_MON_RING_MASK_0,
 		ATH12K_RX_MON_RING_MASK_1,
 		ATH12K_RX_MON_RING_MASK_2,
+		0,
 	},
 	.rx = {
 		0, 0, 0, 0,
@@ -321,27 +323,32 @@ const struct ath12k_hw_ring_mask ath12k_
 		ATH12K_RX_RING_MASK_1,
 		ATH12K_RX_RING_MASK_2,
 		ATH12K_RX_RING_MASK_3,
+		0, 0, 0, 0,
 	},
 	.rx_err = {
 		0, 0, 0,
 		ATH12K_RX_ERR_RING_MASK_0,
+		0, 0, 0, 0, 0, 0, 0, 0,
 	},
 	.rx_wbm_rel = {
 		0, 0, 0,
 		ATH12K_RX_WBM_REL_RING_MASK_0,
+		0, 0, 0, 0, 0, 0, 0, 0,
 	},
 	.reo_status = {
 		0, 0, 0,
 		ATH12K_REO_STATUS_RING_MASK_0,
+		0, 0, 0, 0, 0, 0, 0, 0,
 	},
 	.host2rxdma = {
 		0, 0, 0,
 		ATH12K_HOST2RXDMA_RING_MASK_0,
+		0, 0, 0, 0, 0, 0, 0, 0,
 	},
 	.tx_mon_dest = {
 		ATH12K_TX_MON_RING_MASK_0,
 		ATH12K_TX_MON_RING_MASK_1,
-		0, 0, 0, 0, 0, 0
+		0, 0, 0, 0, 0, 0, 0, 0, 0, 0
 	},
 };
 
--- a/drivers/net/wireless/ath/ath12k/hw.h
+++ b/drivers/net/wireless/ath/ath12k/hw.h
@@ -107,7 +107,7 @@ enum ath12k_bus {
 	ATH12K_BUS_PCI,
 };
 
-#define ATH12K_EXT_IRQ_GRP_NUM_MAX 11
+#define ATH12K_EXT_IRQ_GRP_NUM_MAX 13
 
 struct hal_rx_desc;
 struct hal_tcl_data_cmd;
--- a/drivers/net/wireless/ath/ath12k/pci.c
+++ b/drivers/net/wireless/ath/ath12k/pci.c
@@ -66,12 +66,13 @@ static const struct ath12k_bus_params at
 /* TODO: revisit IRQ mapping for new SRNG's */
 static const struct ath12k_msi_config ath12k_msi_config[] = {
 	{
-		.total_vectors = 16,
+		/* MSI spec expects number of interrupts to be a power of 2 */
+		.total_vectors = 32,
 		.total_users = 3,
 		.users = (struct ath12k_msi_user[]) {
 			{ .name = "MHI", .num_vectors = 3, .base_vector = 0 },
 			{ .name = "CE", .num_vectors = 5, .base_vector = 3 },
-			{ .name = "DP", .num_vectors = 8, .base_vector = 8 },
+			{ .name = "DP", .num_vectors = 12, .base_vector = 8 },
 		},
 	},
 };
